app/controllers/application_controller.rb

```diff
@@ -5,6 +5,149 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
+  end
+
+  get '/' do
+  	erb :index
+  end
+
+  get '/signup' do 
+  	if logged_in?
+  		redirect '/tweets'
+  	else
+		erb :'users/create_user'
+	end
+  end
+
+  post "/signup" do
+    @user = User.create(params)
+
+    if @user.save && params[:username] != "" && params[:password] != ""
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/login' do
+  	if logged_in?
+    	redirect '/tweets'
+    else
+  		erb :'/users/login'
+  	end
+  end
+
+  post "/login" do
+    user = User.find_by(:username => params[:username])
+    
+    if user && user.authenticate(params[:password]) && params[:username] != "" && params[:password] != ""
+        session[:user_id] = user.id
+        redirect "/tweets"
+    else
+        redirect "/failure"
+    end
+
+  end
+
+  get '/users/home' do
+  	if logged_in?
+  		redirect "/users/#{current_user.slug}"
+  	else
+  		redirect '/'
+  	end
+  end
+
+  get '/users/:slug' do
+  		@user = User.find_by_slug(params[:slug])
+  		erb :'/users/show'
+  end
+
+  get '/logout' do
+  	session.clear
+    redirect "/login"
+  end
+
+  get '/tweets' do
+  	if logged_in?
+  		@user = User.find_by_id(session[:user_id])
+  		@tweets = Tweet.all
+  		erb :'/tweets/tweets'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  get '/tweets/new' do
+  	if logged_in?
+  		erb :'tweets/create_tweets'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  post '/tweets/new' do
+  	if params[:content] != ""
+	  	@user = User.find_by_id(session[:user_id])
+	  	@tweet = Tweet.create(params)
+	  	@tweet.user = @user
+	  	@tweet.save
+	  	redirect "/users/#{@user.slug}"
+	 else
+	 	redirect '/tweets/new'
+	 end
+  end
+
+  get '/tweets/:id' do
+  	if logged_in?
+  		@tweet = Tweet.find_by_id(params[:id])
+  		erb :'/tweets/show_tweet'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  get '/tweets/:id/edit' do
+  	if logged_in?
+  		@tweet = Tweet.find_by_id(params[:id])
+  		erb :'/tweets/edit_tweet'
+  	else
+  		redirect '/login'
+  	end
+  end
+
+  patch '/tweets/:id' do
+  	@tweet = Tweet.find_by_id(params[:id])
+  	if params[:content] != ""
+	  	@tweet.content = params[:content]
+	  	@tweet.save
+	  	redirect "/tweets/#{@tweet.id}"
+	else
+		redirect "/tweets/#{@tweet.id}/edit"
+	end
+  end
+
+  delete '/tweets/:id/delete' do
+  	@user = User.find_by_id(session[:user_id])
+  	@tweet = Tweet.find_by_id(params[:id])
+  	if logged_in? && @user.id == @tweet.user_id
```

Comment:
Prefer to find tweets that are already only associated to the `current_user` than to manually check the `user_id` foreign key of the tweet against the `current_user.id`

```ruby
tweet = current_user.tweets.find_by(id: params[:id])
if tweet && tweet.destroy
  redirect "/tweets"
else
  redirect "/tweets/#{tweet.id}"
end
```

We're using the association collection method provided by [`has_many`](http://api.rubyonrails.org/classes/ActiveRecord/Associations/ClassMethods.html#method-i-has_many) to query the user's tweets. 

Note the usage of [`find_by`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find-by) vs [`find`](http://api.rubyonrails.org/classes/ActiveRecord/FinderMethods.html#method-i-find). Had we used `find` here, ActiveRecord would raise an error in the event that we could not find a tweet associated with this user. `find_by` will allow this action to fail gracefully, simply returning nil if the user is trying to delete a tweet that they do not have access to.

Reply:
This one took me a second but I figured it out in the end. 
I just had to change the redirect path on the example you put from "/tweets/#{tweet.id}" to "/tweets/#{params[:id]}". If the tweet variable is nil it wouldn't redirect. Was that an acceptable way to write out the path? I appreciate the code review and have attempted to go over them all to fix/understand them.